### DISCUSSION TITLE
`mon-x` pods get created on a different node to the associated Persistent Volume

### DISCUSSION THREAD

**Discussion by craicoverflow** [rook/rook#13586] [Q&A] [answered]
Hi,

I am not sure if this is a bug or something else but can I get some help on this please 🙏🏻 

When I create my Ceph cluster, the `rook-ceph-mon-x` canary pods are created along with their PVC/PVs and they are assigned to the same node (`rook-ceph-mon-canary-a` is on the same node as the `rook-ceph-mon-a` PVC etc).

However, the `rook-ceph-mon-a` pod is scheduled onto a different node to `rook-ceph-mon-canary-a`, so it is unable to use the `rook-ceph-mon-a` PVC.

```sh
Multi-Attach error for volume “pvc-499a260-4d4c-46ff-8a65-11c90cbcaa5f” Volume is already exclusively attached to one node and can’t be attached to another
```

Is there something I am doing wrong here?

**Comment by travisn** [comment, marked as answer]
When backing the mons with a PVC, it's expected that the corresponding PV would have a topology label or property that constrains where a pod could be scheduled. Therefore, Rook doesn't try to constrain the mon pod to a specific node since the volume should restrict it. In cloud environments, this allows mons to move between nodes, for example nodes within the same AZ. 

If you can't control the PV topology property, a workaround is to label which nodes you expect the mons to run on, and add a `placement.mon.nodeAffinity` for the mons to run on those nodes. See [this example](https://github.com/rook/rook/blob/release-1.13/deploy/examples/cluster.yaml#L166-L173).

**Reply by craicoverflow** [reply]
@travisn thanks for responding so quickly. At the moment there is no labels added to the PV so I will look to the workaround.

If I have three mon pods (6 when you count the canaries), all spread across three different nodes, how does `placement.mon.nodeAffinity` help in this situation? 

I have three nodes (node-a, node-b, node-c), and I want one mon pod on each.

I am able to limit the mon pods to be scheduled on my three desired nodes with nodeAffinity

```
placement:
  mon:
    nodeAffinity:
      requiredDuringSchedulingIgnoredDuringExecution:
        nodeSelectorTerms:
        - matchExpressions:
          - key: acme.org.com/node-group
            operator: In
            values:
            - ceph-mon-node
```

The mon-x-canary pods are randomly assigned to one of the three nodes which have the `ceph-mon-node` label.

```
node-a: mon-a-canary (pod)
node-b: mon-c-canary (pod)
node-c: mon-b-canary (pod)
```

This assignment means that the associated PVs are on the same node:

```
node-a: mon-a (pv)
node-b: mon-c (pv)
node-c: mon-b (pv)
```

The mon pods are not guaranteed to be on the same node as the PV and they fail with `Multi-Attach error for volume “pvc-499a260-4d4c-46ff-8a65-11c90cbcaa5f” Volume is already exclusively attached to one node and can’t be attached to another`

```
node-a: mon-c (pod)
node-b: mon-b (pod)
node-c: mon-a (pod)
```

Apologies if I am missing something.

**Reply by travisn** [reply]
Seems like you need a `portable: false` mon setting similar to what we have for the OSDs (with `storageClassDeviceSets`). 

Until that is implemented, the only workaround I can think of is to just label one node and start with one mon, then increase to two mons with a second node labeled, etc. But there will still be a problem because if the mons are rescheduled (e.g. a node is drained), it would fail when scheduled to another node.
